pkg/config/parameters.go

```diff
@@ -89,6 +90,82 @@ func (a AccessLogType) Validate() error {
 const EnvoyAccessLog AccessLogType = "envoy"
 const JSONAccessLog AccessLogType = "json"
 
+type AccessLogFields []string
+
+func (a AccessLogFields) Validate() error {
+	// Capture Groups:
+	// Given string "the start time is %START_TIME(%s):3% wow!"
+	//
+	//   0. Whole match "%START_TIME(%s):3%"
+	//   1. Full operator: "START_TIME(%s):3%"
+	//   2. Operator Name: "START_TIME"
+	//   3. Arguments: "(%s)"
+	//   4. Truncation length: ":3"
+	re := regexp.MustCompile(`%(([A-Z_]+)(\([^)]+\)(:[0-9]+)?)?%)?`)
+
+	for key, val := range a.AsFieldMap() {
+		if val == "" {
+			return fmt.Errorf("invalid JSON log field name %s", key)
+		}
+
+		if jsonFields[key] == val {
+			continue
+		}
+
+		// FindAllStringSubmatch will always return a slice with matches
+		// where every slice is a slice of submatches with length of 5 (number of capture groups + 1)
+		tokens := re.FindAllStringSubmatch(val, -1)
+		if len(tokens) == 0 {
+			continue
+		}
+
+		for _, f := range tokens {
+			op := f[2]
+			if op == "" {
+				return fmt.Errorf("invalid JSON field: %s, invalid Envoy format: %s", val, f)
+			}
+
+			_, okSimple := envoySimpleOperators[op]
+			_, okComplex := envoyComplexOperators[op]
+			if !okSimple && !okComplex {
+				return fmt.Errorf("invalid JSON field: %s, invalid Envoy format: %s, invalid Envoy operator: %s", val, f, op)
+			}
+
+			if (op == "REQ" || op == "RESP" || op == "TRAILER") && f[3] == "" {
+				return fmt.Errorf("invalid JSON field: %s, invalid Envoy format: %s, arguments required for operator: %s", val, f, op)
+			}
+
+			if op == "START_TIME" && f[4] != "" {
+				return fmt.Errorf("invalid JSON field: %s, invalid Envoy format: %s, arguments required for operator: %s", val, f, op)
+			}
+			// START_TIME cannot not have truncation length
+		}
+	}
+
+	return nil
+}
+
+func (a AccessLogFields) AsFieldMap() map[string]string {
+	fieldMap := map[string]string{}
+
+	for _, val := range a {
+		parts := strings.SplitN(val, "=", 2)
+		if len(parts) == 1 {
+			operator := ""
+			operator, ok := jsonFields[val]
+			if _, isSimple := envoySimpleOperators[strings.ToUpper(val)]; isSimple && !ok {
+				operator = fmt.Sprintf("%%%s%%", strings.ToUpper(val))
+			}
+
+			fieldMap[val] = operator
```

Comment:
I think we are missing a case here. There's 3 things that can happen
1. The field name exists in `jsonFields`
2. The field name is a case-folding match to a `envoySimpleOperators` entry
3. Neither of the above.

In case (3), we would basically do `fieldMap[val] = ""`. I think that this is OK, because we subsequently validate the mapping. Let's comment this clearly for future readers!

Reply:
This is good feedback thank you. I've reorganized the cases and took another look at the variable names to hopefully clarify what's going on, and of course added some comments to make it extra clear.